### ISSUE TITLE
Feature Request: Add Pixel Unshuffle

### ISSUE THREAD

**Issue by hinthornw** [pytorch/pytorch#2456]
I would like to propose the addition of an inverse operation for the current PixelShuffle function. We have been researching temporally connected feature cascades and have had a lot of success in hierarchically mixing residual information through this function. 

The core operation is essentially the following:
For feature map _fm_ with  batchsize _b_, channels _c_, downsize ratio of _r_, and original height and width of _h,w_, do:

```
out_channel = c*(r**2)
out_h = h//r
out_w = w//r
fm_view = fm.contiguous().view(b, c, out_h, r, out_w, r)
fm_prime = fm_view.permute(0,1,3,5,2,4).contiguous().view(b,out_channel, out_h, out_w)
```

cc @albanD @mruberry @fmassa @vfdev-5

**Comment by soumith** [comment]
@alykhantejani (as the man behind PixelShuffle) should we add this?

**Comment by alykhantejani** [comment]
@hinthornw This sounds like something for a pytorch contrib repo, which is currently in the works to be established. Will ping you back on this once it's set up

**Comment by PiotrDabkowski** [comment]
@alykhantejani Why should unshuffle qualify for the contrib while shuffle for the main repo? 
It is quite weird to implement just the upsample logic without the downsample counterpart. 
Lots of recent work relies on the unshuffle so it would be quite beneficial to put it just next to pixel_shuffle so that people do waste time looking for it.

**Comment by ezyang** [comment]
cc @fmassa

**Comment by fmassa** [comment]
I'd be tempted to add this layer in `torchvision.ops`.

But given that `PixelShuffle` is already in PyTorch, I think we could add it to core PyTorch then, even though it's pretty specific to vision and and be implemented efficiently in a few lines of Python code.

**Comment by Badstu** [comment]
Have you add this function to torch or torchvision or anywhere else? I'm sorry that I can not find docs about PixelUnShuffle.

**Comment by iamhankai** [comment]
Any news?

**Comment by albanD** [comment]
The status is that we would be happy to accept a PR adding this feature!
But not core dev is working on this at the moment.

**Comment by bonlime** [comment]
The implementation from the topic starter is quite efficient already. I haven't seen it being used too often and usually, it's only used once at the beginning of the network. So benefits from adding it to the core are questionable. If you need class you could copy-paste this:
```python
class SpaceToDepth(nn.Module):
    def __init__(self, block_size=4):
        super().__init__()
        assert block_size in {2, 4}, "Space2Depth only supports blocks size = 4 or 2"
        self.block_size = block_size

    def forward(self, x):
        N, C, H, W = x.size()
        S = self.block_size
        x = x.view(N, C, H // S, S, W // S, S)  # (N, C, H//bs, bs, W//bs, bs)
        x = x.permute(0, 3, 5, 1, 2, 4).contiguous()  # (N, bs, bs, C, H//bs, W//bs)
        x = x.view(N, C * S * S, H // S, W // S)  # (N, C*bs^2, H//bs, W//bs)
        return x

    def extra_repr(self):
        return f"block_size={self.block_size}"
```

**Comment by mruberry** [comment]
This has been added by @jbschlosser. See https://pytorch.org/docs/master/generated/torch.nn.PixelUnshuffle.html?highlight=unshuffle#torch.nn.PixelUnshuffle.

**Comment by jkyl** [comment]
Are there plans to add support for exporting this to ONNX, as is possible for pixel_shuffle?

**Comment by mruberry** [comment]
No current plans, @jkyl, but you can file a new issue requesting the feature.
